File: 2024/day01.py
from rich import print


def solve(puzzle):
    left = list()
    right = list()

    for line in puzzle.splitlines():
        loc = line.strip().split()
        left.append(int(loc[0]))
        right.append(int(loc[1]))

    left.sort()
    right.sort()

    part1 = sum(abs(l - r) for l, r in zip(left, right))

    freq = dict()
    for n in right:
        freq[n] = freq.get(n, 0) + 1

    part2 = sum(n * freq.get(n, 0) for n in left)
    return (part1, part2)

File: 2024/src/day01.zig
const std = @import("std");

pub fn solve(allocator: std.mem.Allocator, input: []u8) ![2]u64 {
    var left = std.ArrayList(u32).init(allocator);
    defer left.deinit();

    var right = std.ArrayList(u32).init(allocator);
    defer right.deinit();

    var lines = std.mem.tokenizeAny(u8, input, "\n");
    while (lines.next()) |line| {
        var numbers = std.mem.tokenizeAny(u8, line, " ");
        try left.append(try std.fmt.parseInt(u32, numbers.next().?, 10));
        try right.append(try std.fmt.parseInt(u32, numbers.next().?, 10));
        std.debug.assert(numbers.next() == null);
    }

    std.mem.sort(u32, left.items, {}, std.sort.asc(u32));
    std.mem.sort(u32, right.items, {}, std.sort.asc(u32));

    var sum: u32 = 0;
    for (left.items, right.items) |l, r| {
        sum += @abs(@as(i32, @intCast(l)) - @as(i32, @intCast(r)));
    }

    return [2]u64{ sum, 0 };
}
